File: graph/strongly_connected_components/py/test_scc.py
import scc
import unittest


class SCCTest(unittest.TestCase):
    def testNoEdgeGraph(self):
        g = {0:[],
             1:[],
             2:[],
             3:[],
             4:[]}
        cc = scc.scc(g)
        self.assertNotEqual(cc[0], cc[1])
        self.assertNotEqual(cc[0], cc[2])
        self.assertNotEqual(cc[0], cc[3])
        self.assertNotEqual(cc[0], cc[4])
        self.assertNotEqual(cc[1], cc[2])
        self.assertNotEqual(cc[3], cc[4])

    def testTreeGraph(self):
        g = {0:[1,2],
             1:[3,4],
             2:[5],
             3:[],
             4:[],
             5:[]}
        cc = scc.scc(g)
        self.assertNotEqual(cc[0], cc[1])
        self.assertNotEqual(cc[0], cc[2])
        self.assertNotEqual(cc[0], cc[3])
        self.assertNotEqual(cc[0], cc[4])
        self.assertNotEqual(cc[0], cc[5])
        self.assertNotEqual(cc[1], cc[2])
        self.assertNotEqual(cc[3], cc[4])

    def testGeneralGraph(self):
        g = {0:[],
             1:[0,4],
             2:[1,5],
             3:[1],
             4:[1],
             5:[2],
             6:[4,7],
             7:[5,9],
             8:[6],
             9:[8]}
        cc = scc.scc(g)
        # 0 itself forms a cc.
        # 3 itself forms a cc.
        # 1 and 4 form a cc.
        # 2 and 5 form a cc.
        # 6, 7, 8, and 9 form a cc.
        self.assertEqual(cc[1], cc[4])
        self.assertEqual(cc[2], cc[5])
        self.assertEqual(cc[6], cc[7])
        self.assertEqual(cc[6], cc[8])
        self.assertEqual(cc[6], cc[9])
        # Check different cc are numbered differently.
        self.assertNotEqual(cc[0], cc[1])
        self.assertNotEqual(cc[0], cc[2])
        self.assertNotEqual(cc[0], cc[3])
        self.assertNotEqual(cc[0], cc[6])
        self.assertNotEqual(cc[1], cc[2])
        self.assertNotEqual(cc[1], cc[3])
        self.assertNotEqual(cc[1], cc[6])
        self.assertNotEqual(cc[2], cc[3])
        self.assertNotEqual(cc[2], cc[6])
        self.assertNotEqual(cc[3], cc[6])


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: graph/strongly_connected_components/java/TestSCC.java
import org.junit.*;
import static org.junit.Assert.*;
import static org.hamcrest.Matchers.*;

public class TestSCC {
  @Test
  public void NoEdgeGraph() {
    int[][] graph = {{},{},{},{},{}};
    int[] cc = SCC.scc(graph);
    assertThat(cc[0], not(cc[1]));
    assertThat(cc[0], not(cc[2]));
    assertThat(cc[0], not(cc[3]));
    assertThat(cc[0], not(cc[4]));
    assertThat(cc[1], not(cc[2]));
    assertThat(cc[3], not(cc[4]));
  }

  @Test
  public void TreeGraph() {
    int[][] graph = {{1,2},{3,4},{5},{},{},{}};
    int[] cc = SCC.scc(graph);
    assertThat(cc[0], not(cc[1]));
    assertThat(cc[0], not(cc[2]));
    assertThat(cc[0], not(cc[3]));
    assertThat(cc[0], not(cc[4]));
    assertThat(cc[0], not(cc[5]));
    assertThat(cc[1], not(cc[2]));
    assertThat(cc[3], not(cc[4]));
  }

  @Test
  public void GeneralGraph() {
    int[][] graph = new int[10][];
    graph[0] = new int[]{};
    graph[1] = new int[]{0,4};
    graph[2] = new int[]{1,5};
    graph[3] = new int[]{1};
    graph[4] = new int[]{1};
    graph[5] = new int[]{2};
    graph[6] = new int[]{4,7};
    graph[7] = new int[]{5,9};
    graph[8] = new int[]{6};
    graph[9] = new int[]{8};
    int[] cc = SCC.scc(graph);
    // 0 itself forms a cc.
    // 3 itself forms a cc.
    // 1 and 4 form a cc.
    // 2 and 5 form a cc.
    // 6, 7, 8, and 9 form a cc.
    assertThat(cc[1], is(cc[4]));
    assertThat(cc[2], is(cc[5]));
    assertThat(cc[6], is(cc[7]));
    assertThat(cc[6], is(cc[8]));
    assertThat(cc[6], is(cc[9]));
    // Check different cc are numbered differently.
    assertThat(cc[0], not(cc[1]));
    assertThat(cc[0], not(cc[2]));
    assertThat(cc[0], not(cc[3]));
    assertThat(cc[0], not(cc[6]));
    assertThat(cc[1], not(cc[2]));
    assertThat(cc[1], not(cc[3]));
    assertThat(cc[1], not(cc[6]));
    assertThat(cc[2], not(cc[3]));
    assertThat(cc[2], not(cc[6]));
    assertThat(cc[3], not(cc[6]));
  }

  public static void main(String args[]) {
    org.junit.runner.JUnitCore.main("TestSCC");
  }
}
